DragaliaAPI/Controllers/Dragalia/DragonController.cs

```diff
@@ -0,0 +1,510 @@
+using System.Collections.Immutable;
+using System.Numerics;
+using System.Security;
+using AutoMapper;
+using DragaliaAPI.Database.Entities;
+using DragaliaAPI.Database.Repositories;
+using DragaliaAPI.Database.Utils;
+using DragaliaAPI.Models.Generated;
+using DragaliaAPI.Services;
+using DragaliaAPI.Shared.Definitions;
+using DragaliaAPI.Shared.Definitions.Enums;
+using DragaliaAPI.Shared.Services;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.EntityFrameworkCore;
+using static DragaliaAPI.Database.Utils.DragonConstants;
+
+namespace DragaliaAPI.Controllers.Dragalia;
+
+[Route("dragon")]
+[Consumes("application/octet-stream")]
+[Produces("application/octet-stream")]
+[ApiController]
+public class DragonController : DragaliaControllerBase
+{
+    private readonly IDragonDataService _dragonDataService;
+    private readonly IDragonService dragonService;
+    private readonly IUserDataRepository userDataRepository;
+    private readonly IUnitRepository unitRepository;
+    private readonly IInventoryRepository inventoryRepository;
+    private readonly IStoryRepository storyRepository;
+    private readonly IUpdateDataService updateDataService;
+    private readonly IMapper mapper;
+
+    public DragonController(
+        IUserDataRepository userDataRepository,
+        IUnitRepository unitRepository,
+        IInventoryRepository inventoryRepository,
+        IUpdateDataService updateDataService,
+        IStoryRepository storyRepository,
+        IMapper mapper,
+        IDragonDataService charaDataService,
+        IDragonService dragonService
+    )
+    {
+        this.userDataRepository = userDataRepository;
+        this.unitRepository = unitRepository;
+        this.inventoryRepository = inventoryRepository;
+        this.storyRepository = storyRepository;
+        this.updateDataService = updateDataService;
+        this.mapper = mapper;
+        _dragonDataService = charaDataService;
+        this.dragonService = dragonService;
+    }
+
+    [Route("buildup")]
+    [HttpPost]
+    public async Task<DragaliaResult> Buildup([FromBody] DragonBuildupRequest request)
+    {
+        IEnumerable<Materials> matIds = request.grow_material_list
+            .Where(x => x.type == EntityTypes.Material)
+            .Select(x => x.id)
+            .Cast<Materials>();
+
+        Dictionary<Materials, DbPlayerMaterial> dbMats = await this.inventoryRepository
+            .GetMaterials(this.DeviceAccountId)
+            .Where(
+                dbMat =>
+                    matIds.Contains(dbMat.MaterialId)
+                    || dbMat.MaterialId == Materials.FortifyingDragonscale
+                    || dbMat.MaterialId == Materials.AmplifyingDragonscale
+            )
+            .ToDictionaryAsync(dbMat => dbMat.MaterialId);
+        foreach (GrowMaterialList mat in request.grow_material_list)
+        {
+            if (mat.quantity < 0)
+            {
+                throw new ArgumentException("Invalid quantity for MaterialList");
```

Comment:
You should now throw DragaliaExceptions with an appropriate resultcode which will return a 200 with a bad status_code in data_headers and show a message to the user, should they ever manage to trigger this